cranelift/codegen/src/machinst/buffer.rs

```diff
@@ -1724,11 +1783,65 @@ pub struct MachReloc {
     /// The kind of relocation.
     pub kind: Reloc,
     /// The external symbol / name to which this relocation refers.
-    pub name: ExternalName,
+    pub target: RelocTarget,
     /// The addend to add to the symbol value.
     pub addend: i64,
 }
 
+/// A Relocation target
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+#[cfg_attr(
+    feature = "enable-serde",
+    derive(serde_derive::Serialize, serde_derive::Deserialize)
+)]
+pub enum RelocTarget {
+    /// Points to an [ExternalName] outside the current function.
+    ExternalName(ExternalName),
+    /// Points to a [MachLabel] inside this function.
+    /// This is different from [MachLabelFixup] in that both the relocation and the
+    /// label will be emitted and are only resolved at link time.
+    ///
+    /// There is no reason to prefer this over [MachLabelFixup] unless the ABI requires it.
+    Label(MachLabel),
```

Comment:
A possible idea for this -- previously `MachLabel` was "purely internal" which I felt was a nice property about the type and its usage but this is causing it to "leak" from the compilation unit for public labels. That being said, labels are all fixed by the time a function finishes, so as a possible alternative could `Label(MachLabel)` be replaced with `Func(CodeOffset)`?

That would represent how a relocation can be applied against the function itself, but it's a relative offset from the start of the function. That would avoid the need to export `MachLabel` and keeps it internal and avoids any need to perform lookups/translation from a label to an offset (e.g. no need for `MachLabelSite`). I think that should still be enough to bubble up to everyone processing `MachReloc` since they should have a handle on the location of the function in the final artifact?

Reply:
I don't think we can do that, since in the RISC-V backend we have to create a relocation to this label while the function is still being lowered, and if we just saved the offset it's possible that it could change later right? (I think this is the case, not 100% sure).

I think we could maybe keep `RelocTarget` as is, but rename all the labels that are going to be published while finalizing. That way we keep `MachLabel` internal, and have a separate labeling scheme in `MachLabelFinalized` that is public. I think that could work, but it also means that we would have to:

* Renumber published labels (there shouldn't be many)
* Walk through all relocations and update them if they point to a label.
  * We can do this only if we have at least 1 published label, so we don't slow down the common case.

Reply:
True! My thinking is:

* `AddReloc` takes `ExternalName | MachLabel`, as you've done in this PR
* The concept of a public/private label goes away
* The `name` field of `MachReloc` would become `ExternalName | Func(CodeOffset)` 

Then during `MachBuffer::finish`, when all labels are known, the `MachBuffer::relocs` array is processed. This means that `MachBufferFinalized::relocs` would have a different type or alternatively the `relocs()` method would return `impl Iterator` which takes `MachBuffer::relocs` and then performs a mapping translating a `MachLabel` to a `Func(CodeOffset)`

Reply:
Sounds good, I'll do that